src/engine/BMTournament.php

```diff
@@ -0,0 +1,734 @@
+
+        $this->add_button_array($playerId, $buttonIdArray);
+    }
+
+    public function add_button_array($playerId, $buttonIdArray) {
+        if (!is_array($buttonIdArray)) {
+            throw new LogicException('Each player must have selected a button');
+        }
+
+        foreach ($buttonIdArray as $buttonId) {
+            if (!isset($buttonId)) {
+                continue;
+            }
+
+            if (!is_int($buttonId)) {
+                throw new InvalidArgumentException('Button IDs must be integers');
+            }
+        }
+
+        $this->buttonIdArrayArray[$playerId] = $buttonIdArray;
+    }
+
+    /**
+     * This is a generic caller function that calls each
+     * do_next_step_*() function, based on the current tournament state.
+     */
+    public function do_next_step() {
+        if (!isset($this->tournamentState)) {
+            throw new LogicException('Tournament state must be set.');
+        }
+
+        $funcName = 'do_next_step_'.
+                    strtolower(BMTournamentState::as_string($this->tournamentState));
+        $this->$funcName();
+    }
+
+    /**
+     * This is a generic caller function that calls each
+     * update_tournament_state_*() function, based on the current tournament state.
+     */
+    public function update_tournament_state() {
+        if (!isset($this->tournamentState)) {
+            throw new LogicException('Tournament state must be set.');
+        }
+
+        $funcName = 'update_tournament_state_'.
+                    strtolower(BMTournamentState::as_string($this->tournamentState));
+        $this->$funcName();
+    }
+
+    /**
+     * Perform the logic required at BMTournamentState::START_TOURNAMENT
+     */
+    protected function do_next_step_start_tournament() {
+    }
+
+    /**
+     * Update tournament state from BMTournamentState::START_TOURNAMENT if necessary
+     */
+    protected function update_tournament_state_start_tournament() {
+        $this->tournamentState = BMTournamentState::JOIN_TOURNAMENT;
+    }
+
+    /**
+     * Perform the logic required at BMTournamentState::JOIN_TOURNAMENT
+     */
+    protected function do_next_step_join_tournament() {
+    }
+
+    /**
+     * Update tournament state from BMTournamentState::JOIN_TOURNAMENT if necessary
+     */
+    protected function update_tournament_state_join_tournament() {
+        if (count($this->playerIdArray) < $this->nPlayers) {
+            return;
+        }
+
+        foreach ($this->buttonIdArrayArray as $buttonIdArray) {
+            if (empty($buttonIdArray)) {
+                return;
+            }
+        }
+
+        $this->initialiseRemainCountArray();
+
+        $this->tournamentState = BMTournamentState::SHUFFLE_PLAYERS;
+    }
+
+    /**
+     * Perform the logic required at BMTournamentState::SHUFFLE_PLAYERS
+     */
+    protected function do_next_step_shuffle_players() {
+        if ($this->doShufflePlayers) {
+            shuffle($this->playerIdArray);
```

Comment:
You can't use `shuffle()`, because that's direct randomization, and we never use randomization directly because that breaks replay testing --- that's why we use `bm_rand()` for randomization.

I don't think it's that big a deal to just implement shuffling by iteratively pulling random (`bm_rand()`) elements out of the array; it's ten lines of code, and it would mean we wouldn't have to implement any specialized randomization elements.

Reply:
Okay, I've defined a new function in the bootstrap file called bm_shuffle() that shuffles an array in place via calls to bm_rand().